Fix for tar video download then converted to video format

susered opened the pull request:
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [x] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Explanation of your *pull request* in arbitrary form goes here. Please make sure the description explains the purpose and effect of your *pull request* and is worded well enough to be understood. Provide as much context and examples as possible.


When downloading an embedded Rumble video:

 " -f 'bestvideo[ext=mp4]+bestaudio[ext=m4a]/best[ext=mp4]/best[height<=720]' --merge-output-format mp4 --ffmpeg-location /opt/homebrew/bin/ffmpeg  --recode-video mp4 https://rumble.com/embed/v6xpz4u/?pub=4"

The format is usually a TAR file. The TAR file is downloaded, then converted to the specified video format of the user's choice - in the above example, MP4. If a user, at some point in the future, downloads the exact same forgotten video again, the current youtube-dl implementation has to wait for the download of the TAR file again to determine the MP4, converted from TAR, already exists.

This enhancement/bugfix compares the existing MP4 file name to the to-be-downloaded TAR name+video_id to determine the same file will be downloaded again if conversion to the same video format is requested.

dirkf commented:
Thanks for your proposal.

This problem has not been reported in the issue tracker so the proposed fix cannot be progressed for now. An initial response is that, through either user-specified format selection or built-in extractor behaviour, tar formats should be skipped, as in _yt-dlp_, and so this problem would be fixed with no or only a two-line code change. Any further analysis should be done in a new issue.

Please review #10785 and its linked duplicate issues and then open a new issue for your problem if it's still relevant, following #30839.

bashonly commented:
IIRC from the yt-dlp extractor review, there are query parameters that will return m3u8 formats instead of tar

susered commented:
> m3u8

Interesting response to explore.

I used the Developer mode in my browser for the [https://rumble.com/embed/v6yh2vw/?pub=4](https://rumble.com/embed/v6yh2vw/?pub=4), particularly the Inspector mode. I searched for the m3u8 extension, and downloaded file, [https://rumble.com/hls-vod/6yh2vw/playlist.m3u8](https://rumble.com/hls-vod/6yh2vw/playlist.m3u8). The contents were as follows:

`
#EXTM3U
#EXT-X-VERSION:3
#EXT-X-ALLOW-CACHE:YES
#EXT-X-STREAM-INF:BANDWIDTH=4137000,RESOLUTION=1920x1080
https://1a-1791.com/video/fwe2/74/s8/2/y/o/2/s/yo2sz.haa.tar?r_file=chunklist.m3u8&r_type=application%2Fvnd.apple.mpegurl&r_range=68566016-68567477
#EXT-X-STREAM-INF:BANDWIDTH=2174000,RESOLUTION=1280x720
https://1a-1791.com/video/fwe2/74/s8/2/y/o/2/s/yo2sz.gaa.tar?r_file=chunklist.m3u8&r_type=application%2Fvnd.apple.mpegurl&r_range=36027392-36028849
#EXT-X-STREAM-INF:BANDWIDTH=1084000,RESOLUTION=854x480
https://1a-1791.com/video/fwe2/74/s8/2/y/o/2/s/yo2sz.caa.tar?r_file=chunklist.m3u8&r_type=application%2Fvnd.apple.mpegurl&r_range=17978368-17979819
#EXT-X-STREAM-INF:BANDWIDTH=709000,RESOLUTION=640x360
https://1a-1791.com/video/fwe2/74/s8/2/y/o/2/s/yo2sz.baa.tar?r_file=chunklist.m3u8&r_type=application%2Fvnd.apple.mpegurl&r_range=11755008-11756447
`

All TAR files.

Likewise, with [https://rumble.com/embed/v6ybrm0/?pub=4](https://rumble.com/embed/v6ybrm0/?pub=4). I searched for the m3u8 within the browser's inspector and I get "No matches". This result might be expected because I was just able to download the MP4 file.

dirkf commented:
The _yt-dlp_ extractor adds `ver=2` to the JSON request URL (and adds `/u3/` to the path, but that seems to make no difference). The `ver=2` response contains similar sets of media links in the `.u` and `.ua` members, where the original tarry responses from the no-`ver` response have query params added and are listed under `.tar`, while `.timeline` has a miniature preview MP4, `.audio` has AAC audio, and `.hls` has M3U8.

Basically, we just need to back-port the _yt-dlp_ extractor.

IIRC you can generate a download archive using `--no-download --download-archive ...` with your aleady extracted videos if that should be needed.  

The `ver=2` JSON is also available from the webpage as the value assigned to `v.f['{video_id}']` in the hydration `<script>`.

susered commented:
Well, there is a,

`
--skip-download                      Do not download the video
`

I do not see a, `--no-download`

and the following option seemingly requires knowledge of what is in the TAR file:

`
--download-archive FILE              Download only videos not listed in the
                                     archive file. Record the IDs of all
                                     downloaded videos in it.
`

From the Developer inspector option of [https://rumble.com/embed/v6yh2vw/?pub=4](https://rumble.com/embed/v6yh2vw/?pub=4), I have the following:

`
!function(u,a){var t,r,o,n,i,d,s,c,l,e="Rumble",v={F:0},f=function(){return(new Date).getTime()/1e3};(a=u[e]=u[e]||function(){a._.push(arguments)})._=a._||[],v.f={},v.b={};v.f["v6yh2vw"]={"fps":60,"w":1920,"h":1080,"u":{"tar":{"url":"https:\/\/1a-1791.com\/video\/fwe2\/74\/s8\/2\/y\/o\/2\/s\/yo2sz.caa.tar?r_file=chunklist.m3u8&r_type=application%2Fvnd.apple.mpegurl&r_range=17978368-17979819","meta":{"bitrate":1084,"size":17981440,"w":854,"h":480}},"timeline":{"url":"https:\/\/1a-1791.com\/video\/fwe2\/74\/s8\/2\/y\/o\/2\/s\/yo2sz.Faa.mp4","meta":{"bitrate":10,"size":181562,"w":320,"h":180}},"audio":{"url":"https:\/\/1a-1791.com\/video\/fwe2\/74\/s8\/2\/y\/o\/2\/s\/yo2sz.Gaa.aac","meta":{"bitrate":192,"size":3184641,"w":0,"h":0}},"hls":{"url":"https:\/\/rumble.com\/hls-vod\/6yh2vw\/playlist.m3u8"}},"ua":{"tar":{"360":{"url":"https:\/\/1a-1791.com\/video\/fwe2\/74\/s8\/2\/y\/o\/2\/s\/yo2sz.baa.tar?r_file=chunklist.m3u8&r_type=application%2Fvnd.apple.mpegurl&r_range=11755008-11756447","meta":{"bitrate":709,"size":11765760,"w":640,"h":360}},"480":{"url":"https:\/\/1a-1791.com\/video\/fwe2\/74\/s8\/2\/y\/o\/2\/s\/yo2sz.caa.tar?r_file=chunklist.m3u8&r_type=application%2Fvnd.apple.mpegurl&r_range=17978368-17979819","meta":{"bitrate":1084,"size":17981440,"w":854,"h":480}},"720":{"url":"https:\/\/1a-1791.com\/video\/fwe2\/74\/s8\/2\/y\/o\/2\/s\/yo2sz.gaa.tar?r_file=chunklist.m3u8&r_type=application%2Fvnd.apple.mpegurl&r_range=36027392-36028849","meta":{"bitrate":2174,"size":36034560,"w":1280,"h":720}},"1080":{"url":"https:\/\/1a-1791.com\/video\/fwe2\/74\/s8\/2\/y\/o\/2\/s\/yo2sz.haa.tar?r_file=chunklist.m3u8&r_type=application%2Fvnd.apple.mpegurl&r_range=68566016-68567477","meta":{"bitrate":4137,"size":68577280,"w":1920,"h":1080}}},"timeline":{"180":{"url":"https:\/\/1a-1791.com\/video\/fwe2\/74\/s8\/2\/y\/o\/2\/s\/yo2sz.Faa.mp4","meta":{"bitrate":10,"size":181562,"w":320,"h":180}}},"audio":{"192":{"url":"https:\/\/1a-1791.com\/video\/fwe2\/74\/s8\/2\/y\/o\/2\/s\/yo2sz.Gaa.aac","meta":{"bitrate":192,"size":3184641,"w":0,"h":0}}},"hls":{"auto":{"url":"https:\/\/rumble.com\/hls-vod\/6yh2vw\/playlist.m3u8"}}},"i":"https:\/\/1a-1791.com\/video\/fwe2\/74\/s8\/6\/y\/o\/2\/s\/yo2sz.qR4e.1.jpg","evt":{"v":"\/l\/view...6yh2vw.1dssxck","e":"\/l\/pte...6yh2vw.1mk88sa","wt":1,"t":"\/l\/timeline...6yh2vw.3o.1411m01"},"cc":{"en-auto":{"language":"English (auto)","path":"https:\/\/1a-1791.com\/video\/fwe2\/74\/s8\/11\/y\/o\/2\/s\/yo2sz.fn8Si.vtt"}},"l":"\/v70nqu0-humor-no-but-seriously-why-are-cats-so-determined-to-be-assholes.html","r":1,"title":"Humor: No but seriously why are cats SO determined to be assholes?","author":{"name":"NewsVariable","url":"https:\/\/rumble.com\/user\/NewsVariable"},"player":false,"duration":132,"pubDate":"2025-10-23T01:03:37+00:00","loaded":1,"vid":420700892,"timeline":[0,0],"own":false,"restrict":[-3,0],"autoplay":0,"track":0,"live":0,"live_placeholder":false,"livestream_has_dvr":null,"a":{"timeout":-1,"u":"4","aden":[1,0,0],"ov":false,"ads":[],"a":".4.1unufk.6yh2vw..js.boalxa","ae":".4.1unufk.6yh2vw..js.l9yy5r","ap":[false,0],"loop":[]},"viewer_id":"oLAkb20VuVE",loaded:f()};if(!v.k){var p,m,g,h="https://rumble.com",e="/embedJS/u4",b=function(o,e){return o.opts||(o.opts=[]),v.D(e,function(e,t){var r=o[t];switch(t){case"opts":o[t]=r.concat(e);break;case"ad_sys":o[t]=r?v.M(r,e):e;break;default:o[t]=e}}),o},_=function(){try{return u.self!==u.top}catch(e){return!0}},y="ifr="+(_()?"1":"0")+"&",S=(v.l=f(),u.RumbleErrorHandler||(c=0,l=function(e){var t;return e&&((t=e.match(new RegExp("http[s]?://[^/]*rumble.com(/.*)$")))?t[1]:e)},u.RumbleErrorHandler=function(e){var t,r=e.message,o=e.filename,n=e.lineno,i=e.colno,a=l(o);o==a||r.match(/^Script error\./)||3<++c||(o=document.location+"",r=[l(o),c,r,a,n,i],e.error&&e.error.stack&&r.push(e.error.stack.split("\n").slice(1,3).join("\n")),a="/l/jserr?err="+encodeURIComponent(JSON.stringify(r)),o==r[0]&&(a=h+a),(t=document.createElement("img")).src=a,t.width=t.height=1,t.onload=t.onerror=function(){t.onload=null,t.onerror=null})},u.addEventListener("error",u.RumbleErrorHandler)),[]),P=(v.E=function(e){return e.split("?")[0]},v.B=function(e){var t=e.split("?"),r={};return t&&t[1]&&(t=t[1],new URLSearchParams(t).forEach(function(e,t){r[t]=e})),r},v.C=function(e){var r="";return v.D(e,function(e,t){r+=(r?"&":"")+encodeURIComponent(t)+"="+encodeURIComponent(e)}),r},v.D=function(e,t){var r,o;for(o in e)if(e.hasOwnProperty(o)&&void 0!==(r=t(e[o],o)))return r},v.K=function(e,t){if("undefined"==typeof localStorageBlocked)try{localStorageBlocked="undefined"==typeof localStorage||!localStorage}catch(r){localStorageBlocked=!0}if(void 0===t){if(!localStorageBlocked)try{t=localStorage.getItem(e)}catch(r){localStorageBlocked=!0}return localStorageBlocked?S[e]:parseInt(t)==t?parseInt(t):t}if(S[e]=t,!localStorageBlocked)try{return localStorage.setItem(e,t)}catch(r){localStorageBlocked=!1}return!1},v.L=function(e,t,r,o){if(e)if(!e.match(/^(http[s]?:)?\/\/([^/]*)\//)||r)return(o?"https:":"")+"//"+t+("/"!=e[0]?"/":"")+e;return e},v.M=function(e,t){return e.filter(function(e){return-1!==t.indexOf(e)})},v.I=function(){if("undefined"!=typeof Hls)return!0;var e="hls";try{new URL(document.location.href).searchParams.has("fml")&&(e="hls-dev")}catch(t){}return v.q([e]),!1},function(n){var i,t=.5,r=2,a=5,d=8,s=1,e=0,c=function(e){var t=e.config.liveSyncDuration;e.config.lowLatencyMode=t<=3},o=function(){0},l=function(){var e=i.config.liveSyncDuration;return!!(e&&e<d)&&(i.config.liveSyncDuration=e+s,r<(t*=2)&&(t=r),c(i),!0)},f=function(){o(),a=1<e&&0==a?-1:5,++e,2<=e&&l()&&(e=0)};return function(){var e,t={capLevelToPlayerSize:!1,autoStartLoad:!1,nudgeMaxRetry:10,minAutoBitrate:5e5,fragLoadPolicy:{"default":{maxTimeToFirstByteMs:1e4,maxLoadTimeMs:12e4,timeoutRetry:{maxNumRetry:10,retryDelayMs:0,maxRetryDelayMs:0},errorRetry:{maxNumRetry:10,retryDelayMs:1e3,maxRetryDelayMs:8e3}}}};if(n.meta&&n.meta.hlsJsConfig)for(var r in n.meta.hlsJsConfig)e=n.meta.hlsJsConfig[r],"predictive_disable_iterations"==r?a=e:"min_latency"==r?0:"max_latency"==r?d=e:"latency_interval"==r?s=e:t[r]=n.meta.hlsJsConfig[r];try{new URL(document.location.href).searchParams.has("follow_live_edge")&&(t.maxLiveSyncPlaybackRate=1.25)}catch(o){}return i=new Hls(t),c(i),i.on(Hls.Events.ERROR,function(e,t){t.type!=Hls.ErrorTypes.MEDIA_ERROR||t.details!=Hls.ErrorDetails.BUFFER_STALLED_ERROR&&t.details!=Hls.ErrorDetails.BUFFER_NUDGE_ON_STALL||f()}),i}()}),w=[2,0,1],B=["mp4","webm","hls"],M=function(f,u,p){var g,h=function(){var e=f.v.src||f.v.currentSrc;return e=e.match(/blob:/)&&f.hls_js?f.hls_js.url||e:e},b=(u.get=function(){return e=f,t=u.url,21192597==e.vid.vid?t:(r=v.B(t),o=v.E(t),e.vid.live||(n=0,e.vid.a&&(n=e.vid.a.u||0),r.u=n,r.b=0<e.bandwidth_track?1:0),o+"?"+v.C(r));var e,t,r,o,n},function(e){var t=h(),r=u.get();return f.current_video=u,r!=t||e?r:0});u.check=function(){return!b()},u.play=function(){var e,t;p&&f.hls_js&&!g&&(e=f.hls_js,(t=e.config.liveSyncDuration)&&(e.config.liveSyncDuration=1),e.startLoad(),t)&&(e.config.liveSyncDuration=t),g=!0},u.set=function(e){if(p&&!v.I())return setTimeout(function(){u.set()},100),!1;var t,r,o,n,i,a,d,s=b(e),c=0,l=0;s&&(g=!1,t=f.v,f.res=u.key,0<m&&(f.last_set_url==h()?(c=!t.paused,l=t.currentTime):0<f.video_time&&(l=f.video_time)),l&&!f.vid.live&&(f.ui.s.autoSeek=l),l=!!p&&(l=navigator.userAgent||"",r=/Android/i.test(l),/Chrome/i.test(l)&&!/iPad|iPhone/i.test(l)||r||!v.J),e=t,n=s,i=l,a=u,(o=f).hls_js&&o.hls_media_attached&&((d=o.hls_js).detachMedia(e),d.destroy(),o.hls_js=null),i?(d=o.hls_js=P(a),o.j("hlsJsLoaded",d),d.on(Hls.Events.LEVEL_UPDATED,function(e,t){o.live=t.details.live}),d.loadSource(n),d.attachMedia(e),o.hls_media_attached=1):e.src=n,m++,f.last_set_url=s,t.load(),c)&&(u.play(),t.play())};!function(){if(p){var e=u.meta&&u.meta.rumbleLsConfig,t=v.N(f,u);if(e||t){var r=new URL(u.url);if(e&&(e=/iPhone|iPad/i.test(navigator.userAgent||"")?1:u.meta.rumbleLsConfig.playlistLevel||0,r.searchParams.append("level",e)),t)for(var o in t)r.searchParams.append(o,t[o]);u.url=r.toString()}}}()},x=(v.G=function(r,e){for(var o,t,n={},i=m=0;i<w.length;i++)e[o=B[t=w[i]]]&&(v.A[t]||"hls"==o)&&v.D(e[o],function(e,t){n.hasOwnProperty(t)||(e.key=t,n[t]=e,M(r,n[t],"hls"==o))});return n},v.H=function(e){var r,o,n,i=480;e.res&&(i=e.res),e.vid.live&&!v.J&&v.I(),v.D(e.video_list,function(e,t){n=parseInt(t),"hls"!=r&&("hls"==t&&v.J||0<n&&n<=i&&(!r||r<n)?r=t:(!o||n<o&&0<n)&&(o=t))}),(r=r||o)&&e.video_list[r].set()},v.N=function(e,t){var r,o,n=e.vid.transmit_url_params;if(n){if("object"==typeof sessionStorage)try{var i=sessionStorage.getItem("transmit_session_id");i||(i=v.O(16),sessionStorage.setItem("transmit_session_id",i)),n.session=i}catch(a){}return t&&t.meta&&(r=t.meta.w,o=t.meta.h),n.w=r||e.vid.w||null,n.h=o||e.vid.h||null,n}},v.O=function(e){for(var t="",r="ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_.",o=0;o<e;o++)t+=r.charAt(Math.floor(Math.random()*r.length));return t},v.r=function(){var a={},d={},s={b:function(e,t,r){if("object"!=typeof e){if(a[e]&&!r)return!1;if(a[e]=t=t||1,d[e])for(;o=d[e].pop();)o(e,t);return!0}for(var o in e)s.b(o,e[o],r)},a:function(e,t,r){var o,n,i;for(r=r||{},o=0;i=e[o];o++)a[i]?r[i]=a[i]:(n&&(t=function(t,r){return function(e){s.a([t],r,e)}}(n[0],n[1])),n=[i,t]);n?(d[n[0]]||(d[n[0]]=[]),d[n[0]].push(function(e,t){r[e]=t,n[1](r)})):t(r)}};return s},i=v.r(),d=document,s={},v.s=function(e,t,r){var o,n,i=0;s[e]||(s[e]=1,(o=d.createElement("script")).type="text/javascript",o.src=e,r&&(o.addEventListener("load",function(){r(!0)},!1),o.addEventListener("error",function(){r(!1)},!1)),t&&(o.addEventListener("load",n=function(e){if(i||t())return i=1;e&&setTimeout(n,50)},!1),n(1)),d.head.appendChild(o))},v.q=function(e,t){for(var r,o=0;o<e.length;o++)if("ima3"==e[o])v.s("https://imasdk.googleapis.com/js/sdkloader/ima3.js",function(){return!("undefined"==typeof google||!google||!google.ima||(i.b("ima3"),0))});else if("custom_ui"!=e[o]){r=e[o];v.s(a.s.rd+"/j/pv/1073/"+r+("hls"==r||"hls-dev"==r?"":".r2")+".js",null,function(e){e?t&&t(1):v.s(a.s.rd+"/j/p/"+r+("hls"==r||"hls-dev"==r?"":".r2")+".js?_v="+Date.now(),t)})}},v.c=function(e,t,r){i.a(e,t),r||v.q(e)},a.rl=function(e,t){i.b(e)&&t&&t(v)},[0,1,"position","absolute","relative","fixed","normal","none","auto","width","height","margin","padding","border","display","#FFF","#000","100%","background","opacity"]),C=function(e){return x.hasOwnProperty(e)?x[e]:e},k=function(r,e,o){var n,t;if(!r.style&&r.length)for(t=0;t<r.length;t++)k(r[t],e,o);else v.D(e,function(e,t){n=C(t),o&&""!==r.style[n]||(r.style[n]=C(e))})},H=(v.w=k,v.y=function(e){var r={};return v.D(e,function(e,t){r[C(t)]=C(e)}),r},v.t=function(e,t,r,o,n,i,a,d){o||(o=e,n=t);var s="0",c="0";return d&&d.viewbox_top&&(s=d.viewbox_top),d&&d.viewbox_left&&(c=d.viewbox_left),i=i?" RumbleSVG-"+i:"",a=a||"",0<r.indexOf("stroke")&&(a+="stroke:currentColor;"),[e,t,'<svg style="'+a+'" class="RumbleElm'+i+'" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="'+e+'px" height="'+t+'px" viewBox="'+c+" "+s+" "+o+" "+n+'">'+r+"</svg>"]},v.j=v.r(),v.j.c=function(t,r,o,n){v.j.a([r],function(e){n&&(t=t.querySelectorAll(n)[0]),v.z(t,e[r][2]),o&&("string"==typeof o?t.style.display=o:o.apply(t))})},'<path stroke-width="3" stroke-linejoin="round" d="M19 11L3.2 1.5v19z"/>'),O=(v.j.b({bplay:v.t(32,32,H,19,22,"bplay"),play:v.t(22,22,H,19,22,"play","height:100%;"),loader:v.t(80,10,function(){for(var e,t="<g>",r=0;r<21;r++)t+='<circle cx="'+(10*r+5)+'" cy="5" r="'+(e=(e=r-6)<1||5<e?1:e)+'" />';return t+"</g>"}(),80,10)}),u.requestAnimationFrame?1:0),E={},R=function(){var e=E;E={},g=0,v.D(e,function(e){"function"==typeof e&&e()})},N=(v.a=function(e,t){if(!E[e]&&"function"==typeof t&&(E[e]=t,!g)){if(g=1,!O)return setTimeout(R,1e3/24);u.requestAnimationFrame(R)}},v.i=function(e){var t=v.u("div",0,0,e);return 1<t.childNodes.length?t.childNodes:t.firstChild},v.u=function(e,t,r,o){var n=document.createElement(e);return o&&(n.innerHTML=o),t&&(n.className=t),r&&(n.id=r),G(n),n},v.z=function(e,t){e.innerHTML=t,L(e)},v.y({font:"12px sans-serif",fontWeight:6,lineHeight:6,boxSizing:"content-box",webkitBoxSizing:"content-box",19:1,18:7,11:0,12:0,border:7,9:8,10:8,visibility:"visible",textSizeAdjust:8,textDecoration:7})),G=function(e){var t;(t=e.tagName)&&"path"!=(t=t.toLowerCase())&&"video"!=t&&(k(e,N,!0),"svg"==t?k(e,{fill:"currentColor"},!0):(k(e,{color:"inherit"},!0),L(e)))},L=function(e){var t,r;if(t=e.childNodes)for(r=0;r<t.length;r++)G(t[r])},T=(v.x=function(e,t,r,o){var n="__playerHover";(o=o||e)[n]||(e.addEventListener("mouseout",function(){k(o,o[n][0])}),e.addEventListener("mouseover",function(){k(o,o[n][1])})),o[n]=[r,t],k(o,r)},function(e){var d,n={play:"#fff",scrubber:"#75a642",hover:"#fff",background:"rgba(0, 0, 0, 0.20)",hoverBackground:"rgba(0, 0, 0, 0.4)"},s=this,c=-1,l=(v.D(e,function(e,t){s[t]=e}),s.hasima=1,s.hasInit=0,s.bandwidth_track=0,s.is_pending_n=!1,{}),f=(s.addEvent=function(e,t,r){l[r=r||1]||(l[r]={}),l[r][e]||(l[r][e]=[]),l[r][e].indexOf(t)<0&&l[r][e].push(t);var o="addEvent";e!=o&&s.j(o,e)},s.removeEvent=function(e,t,r){var o;l[r=r||1][e]&&(r&&!t?l[r][e]=[]:(o=l[r][e].indexOf(t),l[r][e].splice(o,1)))},s.hasEventListener=function(r){return v.D(l,function(e,t){if(e[r]&&0<e[r].length)return!0})},s.j=function(r,o,n){var i,a,d=[];return v.D(l,function(e,t){if(e[r]&&(n&&n==t||!n))for(a=e[r],i=0;i<a.length;i++)"function"==typeof o&&(o=o()),d.push(a[i](o))}),d},s.triggerError=function(e,t){s.j("error",{code:e,message:t})},s.l1=function(e,t,r){},s.getSetting=function(e,t){var r,o=!1;return s.vid&&s.vid.player&&s.vid.player[e]&&(r=s.vid.player[e],t&&r[t]&&(o=r[t]),t||(o=r)),o=!o&&n[t]?n[t]:o},s.loadVideoStyles=function(e){s.pp.id="vid_"+s.vid.id;var t,r,o,n,i=!1;function a(e){d&&i!==e&&(d.style.display=e?"block":"none",i=e)}s.vars.opts.title&&s.vid.title&&(d.innerHTML=s.vid.title,d.href=v.L(s.vid.l,"rumble.com"),v.w(d,{outline:0,display:"block",18:"linear-gradient(rgba(0,0,0,.7),rgba(0,0,0,0))",textShadow:"rgba(0,0,0,0.5) 0px 0px 2px",padding:"9px",fontSize:"18px",whiteSpace:"nowrap",overflow:"hidden",textOverflow:"ellipsis",position:"absolute",top:0,left:0,right:0}),i=!0,s.addEvent("ui:animate",function(e){a(!!e.next)}),s.addEvent("ui",function(){a(!!s.ui.shown)}),v.x(d,{textDecoration:"underline"},{textDecoration:"none"})),s.bp&&(t=s.getSetting("colors","play"),r=s.getSetting("colors","hover"),o=s.getSetting("colors","background"),n=s.getSetting("colors","hoverBackground"),v.x(s.bp,{color:r,background:n,borderColor:n},{color:t,background:o,borderColor:o},s.bp_svg),s.bp.style.opacity=1)},s.trackBandwidth=function(e){var t=s.bandwidth_track;(e=s.server.bw_tracking?e:0)!=t&&(s.bandwidth_track=e,s.current_video)&&!s.m&&s.current_video.set()},s.loadVideo=function(e,t,r,o,n){var i=(e="object"!=typeof e?{id:e}:e).id;if(v.b[i]&&(s.vars.playlist||(s.vars.playlist=v.b[i]),i=v.b[i][0]),s.hasInit||(s.hasInit=1,v.c(["ui","custom_ui"],function(){for(var e=0;e<v.d.length;e++)v.d[e](s.id)})),!t)return v.e(i,function(){s.loadVideo(e,1,r,o,n)},{ad_count:s.ad_count||null});if(v.f[i].loaded<9){if(s.triggerError("load","Unable to load video "+i),t)throw"Unable to load video "+i;return 2}v.f[i].cfg=e,v.f[i].plays=0;var a=v.f[i];if(a.restrict&&!f(a.restrict)&&(s.triggerError("restricted","Video is restricted"),v.f[i].restricted=!0,s.j("restricted",i),v.f[i].ua=0),s.vars.opts.region_blocked&&(s.triggerError("regionBlocked","Region is not allowed"),v.f[i].regionBlocked=!0,s.j("regionBlocked",i)),a.id=i,a.is_autoplayed=r,a.experiments=o,a.search_correlator=n,s.vid=a,s.live=2==s.vid.live,s.video_list=0,s.current_video=0,c<0&&(c=(s.vars.opts.noads||s.server.bw_ourads_check&&s.vars.opts.norumbleads)&&a.a?1:0),s.trackBandwidth(a&&a.track||c?1:0),!a.restricted&&a.ua&&(s.video_list=v.G(s,a.ua),v.H(s),s.loadVideoStyles()),v.g(s),s.j("loadVideo",a),v.h(s,1),a.restricted&&t)throw"Video "+i+" is restricted"},function(e){var t,r,o,n,i=document,a=!1;if(!e||e[0]<=-3)return!0;r=e[0],o=e[1];try{t=parent!==u?i.referrer||parent.location.href:i.location.href}catch(d){}if(!t)return parent===u;if(t=t.match(/^https?\:\/\/([^\/?#]+)(?:[\/?#]|$)/i))for(t=t[1].split(".");2<=t.length&&!a;)n=t.join("."),-1<o.indexOf(n)&&(a=!0),t.shift();return r!=a}),t=(s.id?s.id+"-":"")+"Rumble-cls",r="metadata",o={2:4,9:17,10:17,18:16,color:15,clear:"both",overflow:"visible"};(s.vars.opts.minimal||s.vars.opts.nopreload)&&(r="none"),s.vars.quality&&(s.res=parseInt(s.vars.quality)),t=v.i('<div class="'+t+'" allowfullscreen tabindex="-1" style="outline: none;"><div><video muted playsinline hidefocus="hidefocus" style="transition: transform 0.3s ease-in-out;width:100% !important;height:100% !important;display:block" preload="'+r+'"'+(s.vars.opts.cc?' crossorigin="anonymous"':"")+'></video><div style="display:flex;opacity:0" class="bigPlayUI ctp"><div class="bigPlayUIInner ctp" style="display:none"></div></div><a style="display:none" target=_blank rel=noopener></a></div></div>'),v.w(t.firstChild,o),v.w(t,o),v.j.c(t,"bplay","block",".bigPlayUIInner"),s.d.appendChild(t),s.p=t.firstChild,s.pp=t,s.v=s.p.firstChild,function(e){if(!v.A){var t,r="canPlayType",o='video/mp4; codecs="avc1.42E01E',n=[0,o+'"',0,o+', mp4a.40.2"',1,'video/webm; codecs="vp9, vorbis"',2,"application/vnd.apple.mpegurl"],i=[!1,!1,!1];if(!e||!e[r])return;for(t=0;t<n.length;t+=2)e[r](n[t+1])&&(i[n[t]]=!0);v.J=i[2],v.A=i}}(s.v),s.rsz=[0,0],s.bp=s.p.childNodes[1],s.bp_svg=s.bp.childNodes[0],s.hasStyle={},d=s.p.childNodes[2],v.w(s.bp_svg,{fill:"currentColor",9:screen.width<480?"16px":"32px",10:screen.width<480?"16px":"32px",12:screen.width<480?"16px":"32px",cursor:"pointer",borderRadius:"50%"}),v.w(s.bp_svg.firstChild,{height:"auto",width:"100%"}),v.w(s.bp,{display:"flex",opacity:0,position:"absolute",top:0,left:0,width:"100%",height:"100%",cursor:"pointer",alignItems:"center",justifyContent:"center",overflow:"hidden"}),s.v.addEventListener("contextmenu",function(e){return e.preventDefault(),!1}),s.loadVideoStyles()});v.d=[];a.s={rd:h,ru:e,ds:b({opts:["title"]},{"opts":["cc"],"feature_ad_watchtime":true,"watching_now_endpoint":"https:\/\/wn0.rumble.com","show_premium_exclusive_gate":false,"ad_free_end_slot_seconds":270,"pause_ads":{"show_pause_ads":true,"pause_ads_time_ms":4000,"pause_ads_sdk":"https:\/\/a.ads.rmbl.ws\/pausejs?a=469"},"do_not_add_to_watch_history":false}),rp:h+e+"/?"+y+"dref=&request=",server:{"bw_tracking":1,"bw_noads_check":1}};v.k={},a.gdpr=2,v.m=function(e,t,r,o){var n=new XMLHttpRequest;n.onreadystatechange=function(){4==n.readyState&&200==n.status&&t(JSON.parse(n.responseText))},n.open("GET",(o?"":a.s.rp)+e),n.send()},v.e=function(e,t,r){var o,n,i,a=[];for("object"!=typeof e&&(e=[e]),o=0;o<e.length;o++)n=e[o],(!v.f[n]||1<v.f[n].loaded&&v.f[n].loaded+(1==e.length?900:1800)<f())&&(v.f[n]={loaded:0,wait:[]}),0==(i=v.f[n]).loaded&&(a.push(n),i.loaded=1),t&&i.loaded<9&&i.wait.push(t);return 0<a.length?(r=r?"&ext="+encodeURIComponent(JSON.stringify(r)):"",r=(r+="")+"&ad_wt="+(v.K("ad_wt")||0),v.m("video&ver=2&v="+a[0]+r,function(e){var t,r,o=[],n={};for(e.loaded||!e?n[a[0]]=e:n=e,v.D(n,function(e,t){for(;r=v.f[t].wait.pop();)o.indexOf(r)<0&&o.push(r);e&&(v.f[t]=e,v.f[t].loaded=f())}),t=0;t<o.length;t++)o[t]()}),1):(t&&t(),0)},a.resize=function(){v.D(v.k,function(e){v.h(e)})},v.h=function(e,t){var r,o,n=!e.rsz,i=[e.pp.clientWidth,e.pp.clientHeight],a=u.innerHeight,d=e.vars;d.resize||(d.resize=_()?"full":"auto");(!n&&(e.rsz[0]!=i[0]||e.rsz[1]!=i[1])||a<i[1]||t)&&(o=d.resize,d.ia&&(o="ia"),e.ui&&e.ui.isFloating&&(o="auto"),r=Math.floor(i[0]/16*9),"ia"==o?screen&&screen.height<r&&(r=screen.height):"full"==o?r=0:"window"==o?r=a:("ctpauto"==o&&e.ui&&e.ui.ctp&&(d.resize="auto"),(a<r||e.ui&&e.ui.getFullScreen())&&(r=0)),"window"!=o&&"ctpauto"!=o&&"auto16:9"!=o&&"full"!=o&&e.vid&&e.vid.a&&e.vid.a.aden&&e.vid.a.aden[2]&&r<360&&0!=r&&!d["float"]&&(r=360),e.rsz[0]!=i[0]&&(n=1),i[1]!=r)&&(n=1,e.pp.style.height=0<r?(i[1]=r)+"px":"100%"),e.rsz&&!n||(e.rsz=i),n&&(v.g(e),e.j("resize"))},v.g=function(e){if(!(!e.vid||e.ui&&e.ui.hasPlayed)){var t,r,o,n=e.vid.i,i=e.vid.t,a=-1,d=e.rsz[0],s=e.rsz[1],c=d/s;if(i)for(t=0;t<i.length;t++)o=c<(o=(r=i[t]).w/r.h)?(s-d/o)*d:(d-s*o)*s,(a<0||o<a)&&(a=o,n=r.i);e.v.poster!=n&&(e.v.poster=n)}},a.$play=function(e,t){var r,o,n=JSON.parse(JSON.stringify(a.s.ds)),i={};if((n=b(n,e)).opts&&(v.D(n.opts,function(e){i[e]=1}),n.opts=i),n.gdpr="undefined"==typeof n.gdpr?2:n.gdpr?1:0,2!=n.gdpr&&(a.gdpr=n.gdpr),v.n=-1==n.analytics||n.opts.minimal?1:0,v.o=n.opts.skip_ga_load?1:0,v.p=n.opts.force_ga_load?1:0,!n.div){if(!p)throw"No div was defined for the player";n.div=p}if(p=o=n.div,!v.k[o]||(r=v.k[o]).d.parentNode||(r=0),n.macros||(n.macros={}),!r){if(!(r=document.getElementById(o))){if(2<t)throw o+" div not found";u.addEventListener("DOMContentLoaded",function(){a.$play(e,3)})}v.k[o]||v.F++,v.k[o]=r=new T({d:r,vid:0,id:o,vars:n,server:a.s.server})}r.loadVideo(n.video),v.h(r)};a.rl("custom_ui");a.$playSettings=function(e){a.s.ds=b(a.s.ds,e)},u.addEventListener("resize",function(){v.a("resize",a.resize)}),u.addEventListener("orientationchange",function(){setTimeout(function(){a.resize()},500)});var D,j,z,I=u.Rumble;for(I._=D=I._||[],D.push=function(e){var t=D.slice.call(e),r=I["$"+t.shift()];"function"==typeof r?r.apply(I,t):t.push.call(D,e)},j=-1,z=D.length;++j<z;)D.push(D.shift())}}(window);
Rumble("play", {"video":{"id":"v6yh2vw"},"div":"player","resize":"full","opts":["force_ga_load"],"viewer_id":"oLAkb20VuVE"});
`

I see a TAR file again:

**v.f["v6yh2vw"]={"fps":60,"w":1920,"h":1080,"u":{"tar":{"url":"https:\/\/1a-1791.com\/video\/fwe2\/74\/s8\/2\/y\/o\/2\/s\/yo2sz.caa.tar?**

Let me know if my interpretation is correct?

Ultimately, I need a backport to resolve my issue?

Thanks again

susered commented:
Just to recap, here is the command I use to download:

`
env PYTHONPATH=/GIT/YouTube-DL/youtube-dl python -m youtube_dl --no-playlist --retries 20 -v --print-traffic -f 'bestvideo[ext=mp4]+bestaudio[ext=m4a]/best[ext=mp4]/best[height<=720]' --merge-output-format mp4 --ffmpeg-location /usr/bin/ffmpeg  --recode-video mp4 https://rumble.com/embed/v6yh2vw/?pub=4
`

If there is an MP4 in the manifest, then it downloads without an _ffmpeg_ conversion. If there is a TAR file, then it downloads, untar, converts to MP4 using _ffmpeg_.

From what I can see, I only have the option to download a TAR file, despite any JSON discovered.

seproDev commented:
Please just take a look at the yt-dlp code: https://github.com/yt-dlp/yt-dlp/blob/master/yt_dlp/extractor/rumble.py

yt-dlp is a fork of youtube-dl and this problem was already solved there months ago. https://github.com/yt-dlp/yt-dlp/issues/12837 / https://github.com/yt-dlp/yt-dlp/issues/12837

You can just download/parse the m3u8. The tar just contains the hls fragments.

susered commented:
Oh! Geesh!!

Thanks ... I get it now.
